buildsystem: clean `dist/tools` with `(dist)clean` as well, parallel cleaning!

crasbe opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md.
-->

### Contribution description

@mguetschow noticed that `make distclean` does not clean `dist/tools`. The cause is described in #21409.

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure

Make sure that the `clean` and `distclean` actually enter the `dist/tools` subdirectory. This can take quite a while.

```
cbuec@W11nMate:~/RIOTstuff/riot-canfix/RIOT$ time make clean --print-directory 2>/dev/null | grep -i dist/to
ols
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/benchmark_udp'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/benchmark_udp'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bootterm'
rm -f /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bootterm/bt
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bootterm'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bossa-1.8'
rm -f /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bossa-1.8/bossac
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bossa-1.8'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bossa-1.9'
rm -f /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bossa-1.9/bossac
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bossa-1.9'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bossa-nrf52'
rm -f /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bossa-nrf52/bossac
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bossa-nrf52'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/cc2538-bsl'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/cc2538-bsl'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/cosy'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/cosy'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/dose'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/dose'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/doxygen'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/doxygen'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/edbg'
rm -f /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/edbg/edbg
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/edbg'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/elf2uf2'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/elf2uf2'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/ethos'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/ethos'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/fixdep'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/fixdep'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/flatc'
rm -rf /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/flatc/bin/build
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/flatc'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/kconfiglib'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/kconfiglib'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/lpc2k_pgm'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/lpc2k_pgm'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/mosquitto_rsmb'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/mosquitto_rsmb'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/mspdebug'
rm -f /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/mspdebug/mspdebug
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/mspdebug'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/nrf52_resetpin_cfg'
rm -rf /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/nrf52_resetpin_cfg/bin/nrf52dk/pkg-build/cmsis
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/nrf52_resetpin_cfg'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/pioasm'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/pioasm'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/riotboot_gen_hdr'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/riotboot_gen_hdr'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/riotboot_serial'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/riotboot_serial'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/setsid'
rm -f /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/setsid/setsid
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/setsid'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/sliptty'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/sliptty'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/teensy-loader-cli'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/teensy-loader-cli'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/tunslip'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/tunslip'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/uf2'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/uf2'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/uhcpd'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/uhcpd'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/zep_dispatch'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/zep_dispatch'

real    4m31.670s
user    1m9.512s
sys     0m18.344s
```

```
cbuec@W11nMate:~/RIOTstuff/riot-canfix/RIOT$ time make distclean --print-directory 2>/dev/null | grep -i dist/tools
make[2]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/doxygen'
make[2]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/doxygen'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/benchmark_udp'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/benchmark_udp'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bootterm'
rm -f /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bootterm/bt
rm -rf /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bootterm/bin
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bootterm'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bossa-1.8'
rm -f /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bossa-1.8/bossac
rm -rf /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bossa-1.8/bin
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bossa-1.8'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bossa-1.9'
rm -f /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bossa-1.9/bossac
rm -rf /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bossa-1.9/bin
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bossa-1.9'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bossa-nrf52'
rm -f /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bossa-nrf52/bossac
rm -rf /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bossa-nrf52/bin
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/bossa-nrf52'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/cc2538-bsl'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/cc2538-bsl'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/cosy'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/cosy'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/dose'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/dose'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/doxygen'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/doxygen'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/edbg'
rm -f /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/edbg/edbg
rm -rf /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/edbg/bin
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/edbg'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/elf2uf2'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/elf2uf2'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/ethos'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/ethos'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/fixdep'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/fixdep'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/flatc'
rm -rf /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/flatc/bin/build
rm -rf /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/flatc/bin
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/flatc'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/kconfiglib'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/kconfiglib'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/lpc2k_pgm'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/lpc2k_pgm'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/mosquitto_rsmb'
rm -rf /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/mosquitto_rsmb/bin
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/mosquitto_rsmb'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/mspdebug'
rm -f /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/mspdebug/mspdebug
rm -rf /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/mspdebug/bin
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/mspdebug'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/nrf52_resetpin_cfg'
rm -rf /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/nrf52_resetpin_cfg/bin/nrf52dk/pkg-build/cmsis
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/nrf52_resetpin_cfg'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/pioasm'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/pioasm'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/riotboot_gen_hdr'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/riotboot_gen_hdr'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/riotboot_serial'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/riotboot_serial'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/setsid'
rm -f /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/setsid/setsid
rm -rf /home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/setsid/bin
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/setsid'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/sliptty'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/sliptty'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/teensy-loader-cli'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/teensy-loader-cli'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/tunslip'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/tunslip'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/uf2'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/uf2'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/uhcpd'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/uhcpd'
make[1]: Entering directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/zep_dispatch'
make[1]: Leaving directory '/home/cbuec/RIOTstuff/riot-canfix/RIOT/dist/tools/zep_dispatch'

real    3m43.434s
user    1m5.344s
sys     0m15.711s
```

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references

Fixes #21409.

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

mguetschow commented:
Thanks a lot for taking care of this! I don't have time to test this today, but I'd like to double-check it also works with an out-of-tree `RIOTBUILD` folder.

crasbe commented:
I couldn't help it... `make clean` and `make distclean` take soooooooooo long and I remembered that there was this `ALL--` thing in our Make system to do some parallel building...

Well, the speed difference is quite significant:
```
cbuec@W11nMate:~/RIOTstuff/riot-canfix/RIOT$ git checkout pr/distclean_dist
Switched to branch 'pr/distclean_dist'
cbuec@W11nMate:~/RIOTstuff/riot-canfix/RIOT$ time make distclean -j12 &>/dev/null

real    0m27.681s
user    1m50.061s
sys     0m36.670s
cbuec@W11nMate:~/RIOTstuff/riot-canfix/RIOT$ time make clean -j12 &>/dev/null

real    0m29.072s
user    1m47.513s
sys     0m38.500s

cbuec@W11nMate:~/RIOTstuff/riot-canfix/RIOT$ git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
cbuec@W11nMate:~/RIOTstuff/riot-canfix/RIOT$ time make distclean -j12 &>/dev/null

real    4m33.860s
user    1m11.763s
sys     0m21.099s
```

crasbe commented:
The `dist/tools/*: add RIOTBASE and distclean to Makefiles` commit is optional, with the `-` prefix in the recipes, Make will ignore errors that occur in the submakes.

mguetschow commented on makefiles/app_dirs.inc.mk:
```diff
@@ -16,6 +17,12 @@ APPLICATION_DIRS := $(shell find $(APPS_BASE_DIRS) -name Makefile -type f | \
+# used for `make claen` and `make distclean`
+TOOLS_DIRS := $(shell find $(TOOLS_BASE_DIRS) -mindepth 2 -maxdepth 2 -name Makefile -type f | \
+    xargs dirname | \
+    grep -vFf $(RIOTBASE)/makefiles/app_dirs.blacklist | \
```
Pretty sure this wouldn't filter out anything here, would it?

crasbe replied:
No, I put it in as an option to filter out tools that shouldn't be cleaned. It was an alternative to the commit that fixes the "broken" Makefiles.

It doesn't do anything, but it won't hurt anything either.
